### PR TITLE
Support HyperV Passthrough: automatically use all available HyperV features

### PR DESCRIPTION
**Pull request by iholder101**
### What this PR does
Before this PR:
In order to use HyperV features, the VM creator had to enable these features explicitly.

After this PR:
The VM owner can choose to use HyperV passthrough which will automatically use all available HyperV features on the node, even if libvirt is not aware of them.

Related links:
- https://bugzilla.redhat.com/show_bug.cgi?id=1851249

### Why we need it and why it was done in this way
The following tradeoffs were made:
There is a clear trade-off when using hyperv-passthrough.

On the one hand, the admin does not have to worry about which features he needs to enable/disable, which nodes support these features, etc. Features that are not yet supported in libvirt would also be available.
On the second hand, the VM would not be able to migrate as it is dangerous. As written in libvirt docs: [1]

```Migration of a guest using passthrough is dangerous if the source and destination hosts are not identical in both hardware, QEMU version, microcode version and configuration. If such a migration is attempted then the guest may hang or crash upon resuming execution on the destination host.```

[1] https://libvirt.org/formatdomain.html#hypervisor-features

The following alternatives were considered:
I've chose to create two "distinct features" at the API level: HyeprV (that includes the different features) and HyperVPassthrough. Retrospectively, having a HyperV struct that contains both a mode knob (mode=passthrough for example) and a list of features would be ideal. However, currently this will break backward compatibility. Therefore, the user can either choose to enable HyperV alongside his needed features, or to enable HyperVPassthrough without an ability to explicitly enable/disable features.

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] ~Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required~
- [X] PR: The PR description is expressive enough and will help future contributors
- [X] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [X] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [X] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [X] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] ~Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered~

### Release note

```release-note
Support HyperV Passthrough: automatically use all available HyperV features
```

### CODE REVIEW

File: tests/vmi_configuration_test.go [hunk 1]
```diff
@@ -1732,6 +1735,23 @@ var _ = Describe("[sig-compute]Configurations", decorators.SigCompute, func() {
 				Expect(origVmiWithoutHeadroom.Spec.Domain.Memory).To(Equal(vmiWithoutHeadroom.Spec.Domain.Memory), "vmi guest memory is not expected to change")
 			})
 		})
+
+		FIt("be able to use hyperv passthrough and ensure it's non-migratable", func() {
```
Comment (enp0s3) [inline tests/vmi_configuration_test.go:1739]: nit: `FIt` commited

Comment (iholder101) [reply tests/vmi_configuration_test.go]: Thank you!
Fixed :+1:

**Review by enp0s3 — COMMENTED** [review]
@iholder101 Hey! can we re-use 
```
spec.domain.cpu.model: host-passthrough`
```
For that occasion? or is this a different scenario?

File: pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter_test.go [hunk 2]
```diff
@@ -4754,6 +4754,50 @@ var _ = Describe("Validating VMICreate Admitter", func() {
...
+				Request: &admissionv1.AdmissionRequest{
+					Resource: webhooks.VirtualMachineInstanceGroupVersionResource,
+					Object: runtime.RawExtension{
+						Raw: vmiBytes,
+					},
+				},
+			}
+			resp := vmiCreateAdmitter.Admit(ar)
+
+			//causes := ValidateVirtualMachineInstanceSpec(k8sfield.NewPath("fake"), &vmi.Spec, config)
```
Comment (fossedihelm) [inline pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter_test.go:4785]: leftover

File: tests/libvmi/cpu.go [hunk 3]
```diff
@@ -69,3 +71,15 @@ func WithNUMAGuestMappingPassthrough() Option {
...
+	return func(vmi *v1.VirtualMachineInstance) {
+		if vmi.Spec.Domain.Features == nil {
+			vmi.Spec.Domain.Features = &v1.Features{}
+		}
+		if vmi.Spec.Domain.Features.HypervPassthrough == nil {
+			vmi.Spec.Domain.Features.HypervPassthrough = &v1.HyperVPassthrough{}
+		}
+		vmi.Spec.Domain.Features.HypervPassthrough.Enabled = virtpointer.P(true)
+	}
+}
```
Comment (fossedihelm) [inline tests/libvmi/cpu.go:85]: I am not sure this is located in the right file :)

File: tests/vmi_configuration_test.go [hunk 4]
```diff
@@ -1732,6 +1735,23 @@ var _ = Describe("[sig-compute]Configurations", decorators.SigCompute, func() {
...
+			Expect(err).ToNot(HaveOccurred())
+
+			domSpec := &api.DomainSpec{}
+			Expect(xml.Unmarshal([]byte(domXml), domSpec)).To(Succeed())
+			Expect(domSpec.Features.Hyperv.Mode).To(Equal(api.HypervModePassthrough))
+
+			conditionManager := controller.NewVirtualMachineInstanceConditionManager()
+			cond := conditionManager.GetCondition(vmi, v1.VirtualMachineInstanceIsMigratable)
+			Expect(cond).ToNot(BeNil())
+			Expect(cond.Status).To(Equal(kubev1.ConditionFalse))
```
Comment (fossedihelm) [inline tests/vmi_configuration_test.go:1753]: This can be flaky:
```suggestion

			Eventually(matcher.ThisVMI(vmi), 60*time.Second, 1*time.Second).Should(matcher.HaveConditionFalse(v1.VirtualMachineInstanceIsMigratable))
```

File: tests/libvmi/cpu.go (hunk 3 above)
Comment (fossedihelm) [reply tests/libvmi/cpu.go:85]: Maybe just create a function inside hyperv_test.go file as per `withReEnlightenment`?

File: tests/vmi_configuration_test.go [hunk 5]
```diff
@@ -1732,6 +1735,23 @@ var _ = Describe("[sig-compute]Configurations", decorators.SigCompute, func() {
 				Expect(origVmiWithoutHeadroom.Spec.Domain.Memory).To(Equal(vmiWithoutHeadroom.Spec.Domain.Memory), "vmi guest memory is not expected to change")
 			})
 		})
+
+		It("be able to use hyperv passthrough and ensure it's non-migratable", func() {
```
Comment (fossedihelm) [inline tests/vmi_configuration_test.go:1739]: Should this be moved inside hyperv_test file?

**Review by fossedihelm — COMMENTED** [review]
Hey @iholder101! Thank you for this effort! Some comments below :)

**Comment by iholder101** [comment]
> @iholder101 Hey! can we re-use
> 
> ```
> spec.domain.cpu.model: host-passthrough`
> ```
> 
> For that occasion? or is this a different scenario?

Hey @enp0s3!
`spec.domain.cpu.model` determines the CPU model (e.g. Skylake, Cannonlake, etc) and its features (e.g. SVM/VTx, etc).

However, HyperV features are different.
When you enable a Hyper-V feature like an enlightenment, it allows the guest operating system (the OS running inside the virtual machine) to be aware that it’s running on a virtualized hardware and not on a physical machine. This awareness allows the guest OS to optimize its operations for a virtual environment, which can lead to improved performance.

So TLDR: these are distinct controls for different things.

@vladikr to keep me honest here.

File: pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter_test.go (hunk 2 above)
Comment (iholder101) [reply pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter_test.go]: Removed, thanks!

File: tests/vmi_configuration_test.go (hunk 4 above)
Comment (iholder101) [reply tests/vmi_configuration_test.go]: Awesome! Thanks a lot!
done

File: tests/libvmi/cpu.go (hunk 3 above)
Comment (iholder101) [reply tests/libvmi/cpu.go]: Yeah, we should introduce these hypervisor features into libvmi, but you're right, cpu is not the ideal pleace for it.
We can handle this in a follow-up :) thanks!

> Maybe just create a function inside hyperv_test.go file as per withReEnlightenment?

Done!

File: tests/vmi_configuration_test.go (hunk 5 above)
Comment (iholder101) [reply tests/vmi_configuration_test.go]: Yeah I was conflicting on it as well, but I think you're right.

I've made a small refactoring for the hyperv file. Along the way I've removed the `Serial` label from more than half of the tests. So your feedback actually helped improve CI performance in general :) thank you!

**Review by fossedihelm — COMMENTED** [review]
Thank you @iholder101!
We are almost there! Some small nits :)

File: tests/vmi_configuration_test.go (hunk 5 above)
Comment (fossedihelm) [reply tests/vmi_configuration_test.go]: :slightly_smiling_face:

File: tests/hyperv_test.go [hunk 6]
```diff
@@ -278,6 +283,22 @@ var _ = Describe("[sig-compute] Hyper-V enlightenments", decorators.SigCompute,
 			Entry("Verify that features aren't applied when enabled is false", &v1.FeatureState{Enabled: pointer.BoolPtr(false)}),
 		)
 	})
+
+	Context("VMI with HyperV passthrough", func() {
+		It("be able to use hyperv passthrough and ensure it's non-migratable", func() {
```
[the remark this answers is not in the document]
Comment (iholder101) [reply tests/hyperv_test.go]: Thanks! :+1:

**Comment by fossedihelm** [comment]
Looks good! Nothing else from my side
/lgtm

**Review by enp0s3 — COMMENTED** [review]
/approve
@iholder101 Thank you!

**Review by EdDev — COMMENTED** [review]
I am a bit surprised.

This change includes API additions but I did not manage to find any ref of a design proposal discussing it.
It manage to get merged in just a few days, making it hard to open it for a wider discussion.

**Comment by fabiand** [comment]
I'm not sure if a design is really needed here, as in effect this is just about exposing one more low-level hyperv flag of qemu.

**Comment by EdDev** [comment]
> I'm not sure if a design is really needed here, as in effect this is just about exposing one more low-level hyperv flag of qemu.

I guess we do not have a policy about this, but API changes are impactful, especially when they may have dependency implications with other values.
I do not know the details, I just ask to keep such changes open to allow more visibility and feedback.

**Comment by enp0s3** [comment]
@EdDev I see that there is a design proposal for [API review](https://github.com/kubevirt/community/pull/221) but it looks like its still open for discussions. Perhaps we need to think about a fast-track for API graduation for fields that are urgently needed i.e. using a sidecar annotation that will inject it to domain XML, and at the same time starting an API enhancement process in upstream.

**Comment by fabiand** [comment]
> API changes are impactful, 

You know they are! And thanks for caring for this case. In that sense you are right: It's an api change, and this qas quickly merged.

It's secondary that it is a) small b) backwards compatible and c) design impact.

If I hear you right, then I agree: because it was an api change. It should have seen more attention.

I've spoken to @dhiller  and @vladikr about this in the past week. One question is: For anything that is touching the API, can we require approval from 2 approvers?
It seems like prow does not permit this right now.

@EdDev this this (2+ approvers required) something that you'd consider valueable for API changes. and yes: do you know how to change prow plugins? (hint, hint)

**Comment by enp0s3** [comment]
@fabiand Hey. I think it's not only about the number of approvers, the PR should be approved by each SIG it is touching. For example for this PR I would expect at least 3 members to approve it: 
* sig-tests because this PR adds a new e2e test
* sig-api due to API change
* sig-compute

**Comment by iholder101** [comment]
> I am a bit surprised.
> 
> This change includes API additions but I did not manage to find any ref of a design proposal discussing it.
> It manage to get merged in just a few days, making it hard to open it for a wider discussion.

While I agree that a) API changes are very important and b) that we need a better more strict process of reviewing and approving them, I don't understand your confusion as today we don't have a clear policy for that and API changes are being regularly merged without a design proposal. You kinda make it sound as if it's obvious that every API change requires a design proposal which is not true (yet :smile:).

I any case, I think that this specific case was easier than others. As @fabiand said, the new field is straight-forward in a lot of ways. Not only is it backward compatible and small, but it also propagates a flag to QEMU that says "configure hyperv for me". This means that the HyperV API cannot be used alongside it (which is covered by this PR), and that the API is not expandible in any significant way.

As to the general case:
>One question is: For anything that is touching the API, can we require approval from 2 approvers?

> Hey. I think it's not only about the number of approvers, the PR should be approved by each SIG it is touching.

Yes! Huge +1.
IMHO, optimally such a PR would need an approval from a few SIGs - sig-api, sig-compute, and perhaps sig-testing as well. This way every reviewer focuses on a specific area and approves whatever he's expert on. A similar approach is taken by Kubernetes btw.

**Comment by EdDev** [comment]
> @EdDev this this (2+ approvers required) something that you'd consider valueable for API changes. and yes: do you know how to change prow plugins? (hint, hint)

I favor @enp0s3 [suggestion](https://github.com/kubevirt/kubevirt/pull/11500#issuecomment-2002358686) .

The problem to implement it is the approvers structure at the moment (all root approvers), but this can be "fixed" on a per-SIG case and prow supports it. The OWNERS files can be configured to override any previous OWNERS files in the hierarchy.
The project maintainers can agree to assign two super-power maintainers that will appear in all OWNERS files and these maintainers will use their power with great care, in emergency or critical situations.

> While I agree that a) API changes are very important and b) that we need a better more strict process of reviewing and approving them, I don't understand your confusion as today we don't have a clear policy for that and API changes are being regularly merged without a design proposal. You kinda make it sound as if it's obvious that every API change requires a design proposal which is not true (yet 😄).

I find this change as a good example of why it is needed (design proposal) on any API change.
API changes, IMO, should have hard time getting merged in a stable production project like ours.

> I any case, I think that this specific case was easier than others. As @fabiand said, the new field is straight-forward in a lot of ways. Not only is it backward compatible and small, but it also propagates a flag to QEMU that says "configure hyperv for me". This means that the HyperV API cannot be used alongside it (which is covered by this PR), and that the API is not expandible in any significant way.

This may be indeed correct, however, it does not consider other opinions or alternatives.
It lacks visibility and opportunity for discussion and review.
In this case, it even got missed by the `sig-api`.
 
> > Hey. I think it's not only about the number of approvers, the PR should be approved by each SIG it is touching.
> 
> Yes! Huge +1. IMHO, optimally such a PR would need an approval from a few SIGs - sig-api, sig-compute, and perhaps sig-testing as well. This way every reviewer focuses on a specific area and approves whatever he's expert on. A similar approach is taken by Kubernetes btw.

+1
